external/llvm-project/mlir/lib/Dialect/Async/IR/Async.cpp

```diff
@@ -55,11 +55,177 @@ YieldOp::getMutableSuccessorOperands(Optional<unsigned> index) {
+          result.attributes)) {
+    return ::mlir::failure();
+  }
+
+  // Parse dependency tokens.
+  int32_t numDependencies = 0;
+  if (succeeded(parser.parseOptionalLSquare())) {
+    SmallVector<OpAsmParser::OperandType, 4> tokenArgs;
+    if (parser.parseOperandList(tokenArgs) ||
+        parser.resolveOperands(tokenArgs, tokenTy, result.operands) ||
+        parser.parseRSquare())
+      return failure();
+
+    numDependencies = tokenArgs.size();
+  }
+
+  if (parser.parseLParen())
+    return failure();
+  operandsOperandsLoc = parser.getCurrentLocation();
+  if (parser.parseOperandList(operandsOperands))
+    return failure();
+  if (parser.parseRParen())
+    return failure();
+
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+  if (parser.parseColon())
+    return failure();
+
+  FunctionType operands__allResult_functionType;
+  if (parser.parseType(operands__allResult_functionType))
+    return failure();
+  operandsTypes = operands__allResult_functionType.getInputs();
+  auto resultTypes = operands__allResult_functionType.getResults();
+  allResultTypes.append(resultTypes.begin(), resultTypes.end());
+  result.addTypes(allResultTypes);
+  if (parser.resolveOperands(operandsOperands, operandsTypes,
+                             operandsOperandsLoc, result.operands))
+    return failure();
+
+  // Add derived `operand_segment_sizes` attribute based on parsed operands.
+  int32_t numOperands = result.operands.size() - numDependencies;
+  auto operandSegmentSizes =
+      DenseIntElementsAttr::get(VectorType::get({2}, IntegerType::get(ctx, 32)),
+                                {numDependencies, numOperands});
+  result.addAttribute(kOperandSegmentSizesAttr, operandSegmentSizes);
+
+  return success();
+}
+
+LogicalResult LaunchOp::verify() {
+  MLIRContext *ctx = getContext();
+  auto tokenTy = TokenType::get(ctx);
+
+  // The 'callable' must be a kernel and resolved.
+  CallOpInterface callIf(*this);
+  auto *callable = callIf.resolveCallable();
+  if (!callable)
+    return emitOpError("requires a resolved callable");
+  FuncOp func = dyn_cast<FuncOp>(callable);
+
+  if (!func || !func->hasAttr("kernel"))
+    return emitOpError("requires a 'kernel' func reference");
+
+  auto funcType = func.getType();
+  // The result types should be a leading async.token and matching return types
+  // of the kernel func.
+  auto resultTypes = getResultTypes();
+  if (resultTypes.size() != (funcType.getResults().size() + 1))
+    return emitOpError(
+        "requires matching result types with a leading async.token");
+
+  auto resultItr = ++resultTypes.begin();
+  for (auto resType : funcType.getResults()) {
+    if (*resultItr++ != resType)
+      return emitOpError("requires matching result types with func");
+  }
+
+  // The dependencies must be async.tokens
+  for (auto dep : dependencies()) {
+    if (dep.getType() != tokenTy)
+      return emitOpError("requires all dependencies to be async.token");
+  }
```

Comment:
Seems like this should be part of a declarative type constraint?

Reply:
It is mostly (see $dependencies in AsyncOps.td) but they are co-mingled with operands at the Op level and dis-ambiguated with operand_segment_sizes. There may be ways around the creation to mess it up, so this is a little sanity.

Reply:
This is completely redundant with the code that'll get automatically generated by the type constraints in the tablegen, which does pretty much this loop

Reply:
I did add a somewhat generic clone method on the CallOpInterface, which can be used by anyone and with any new operands, so I think it is possible to get a bad op.

Reply:
The fact that LaunchOp is defined as taking `Variadic<Async_TokenType>` means that the code that checks to make sure all the operands to the first variadic argument (the one with the tokens) is an `Async_TokenType` will be generated by tablegen and will run in addition to this verify(). You don't need to include the verification code again here.

Reply:
No, cloning does not go thru that code. It adds operands directly.

Reply:
But why is there cloning? And ::verify() is ::verify() and will check all invariants, not just the user-specified ones. I'm pretty sure you don't need this.